File: framework/languages/java/e2e/PubSub/Server/Subscriber/src/main/java/systems/zlink/e2e/pubsub/subscriber/Infrastructure/FanoutObserverController.java
package Infrastructure;

import systems.zlink.e2e.pubsub.subscriber.Infrastructure;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Flow;
import java.util.concurrent.atomic.AtomicBoolean;
import org.springframework.beans.factory.ObjectProvider;
import systems.zlink.e2e.pubsub.shared.Contracts;
import systems.zlink.framework.monitoring.ZLinkFanoutStatus;
import systems.zlink.framework.monitoring.ZLinkObservedStatus;
import systems.zlink.framework.runtime.host.ZLinkFrameworkRuntime;

public final class FanoutObserverController {
    private final ObjectProvider<ZLinkFrameworkRuntime> runtime;
    private final Object monitor = new Object();
    private final Map<String, ControlledObserver> observers = new LinkedHashMap<>();
    private final List<Entry> entries = new ArrayList<>();

    public FanoutObserverController(ObjectProvider<ZLinkFrameworkRuntime> runtime) {
        this.runtime = runtime;
    }

    public void start(String name, int capacity, boolean slow) {
        if (name == null || name.isBlank() || capacity <= 0) {
            throw new IllegalArgumentException("observer name and positive capacity are required");
        }
        cancel(name);
        ControlledObserver observer = new ControlledObserver(name, slow);
        synchronized (monitor) { observers.put(name, observer); }
        runtime.getObject().fanoutRuntime()
            .observe(Contracts.EVENT_CHANNEL, capacity).subscribe(observer);
    }

    public void release(String name) {
        synchronized (monitor) {
            ControlledObserver observer = observers.get(name);
            if (observer != null) observer.release();
        }
    }

    public void cancel(String name) {
        synchronized (monitor) {
            ControlledObserver observer = observers.remove(name);
            if (observer != null) observer.cancel();
        }
    }

    public void waitFor(String name, long timeoutMillis) {
        long deadline = System.nanoTime() + timeoutMillis * 1_000_000L;
        synchronized (monitor) {
            while (entries.stream().noneMatch(entry -> entry.observer().equals(name))
                && System.nanoTime() < deadline) {
                try {
                    monitor.wait(Math.max(1L, Math.min(250L,
                        (deadline - System.nanoTime()) / 1_000_000L)));
                } catch (InterruptedException error) {
                    Thread.currentThread().interrupt();
                    throw new IllegalStateException("observer wait interrupted", error);
                }
            }
            if (entries.stream().noneMatch(entry -> entry.observer().equals(name))) {
                throw new IllegalStateException("observer " + name + " did not receive a status");
            }
        }
    }

    public List<Entry> snapshot() {
        synchronized (monitor) { return List.copyOf(entries); }
    }

    public record Entry(
        String observer,
        long sequence,
        String state,
        int readyPublisherCount,
        long lossCoalesced,
        long lossDiscardedTerminal) { }

    private final class ControlledObserver
        implements Flow.Subscriber<ZLinkObservedStatus<ZLinkFanoutStatus>> {
        private final String name;
        private final AtomicBoolean released;
        private Flow.Subscription subscription;

        private ControlledObserver(String name, boolean slow) {
            this.name = name;
            released = new AtomicBoolean(!slow);
        }

        @Override public void onSubscribe(Flow.Subscription value) {
            subscription = value;
            value.request(1);
        }

        @Override public void onNext(ZLinkObservedStatus<ZLinkFanoutStatus> value) {
            var status = value.status();
            synchronized (monitor) {
                entries.add(new Entry(
                    name, status.sequence(), status.state().name(),
                    status.readyPublisherCount(), value.loss().coalescedCount(),
                    value.loss().discardedTerminalCount()));
                monitor.notifyAll();
                while (!released.get()) {
                    try {
                        monitor.wait(250L);
                    } catch (InterruptedException error) {
                        Thread.currentThread().interrupt();
                        return;
                    }
                }
            }
            subscription.request(1);
        }

        @Override public void onError(Throwable error) { synchronized (monitor) { monitor.notifyAll(); } }
        @Override public void onComplete() { synchronized (monitor) { monitor.notifyAll(); } }

        private void release() {
            released.set(true);
            synchronized (monitor) { monitor.notifyAll(); }
        }

        private void cancel() {
            release();
            if (subscription != null) subscription.cancel();
        }
    }
}

File: framework/languages/java/e2e-kotlin/PubSub/Server/Subscriber/src/main/kotlin/systems/zlink/e2e/kotlin/pubsub/subscriber/Infrastructure/FanoutObserverController.kt
package systems.zlink.e2e.kotlin.pubsub.subscriber

import java.util.concurrent.Flow
import java.util.concurrent.atomic.AtomicBoolean
import org.springframework.beans.factory.ObjectProvider
import systems.zlink.e2e.kotlin.pubsub.shared.Contracts
import systems.zlink.framework.monitoring.ZLinkFanoutStatus
import systems.zlink.framework.monitoring.ZLinkObservedStatus
import systems.zlink.framework.runtime.host.ZLinkFrameworkRuntime

data class FanoutObserverEntry(
    val observer: String,
    val sequence: Long,
    val state: String,
    val readyPublisherCount: Int,
    val lossCoalesced: Long,
    val lossDiscardedTerminal: Long,
)

class FanoutObserverController(
    private val runtime: ObjectProvider<ZLinkFrameworkRuntime>,
) {
    private val observers = linkedMapOf<String, ControlledObserver>()
    private val monitor = Object()
    private val entries = mutableListOf<FanoutObserverEntry>()

    fun start(name: String, capacity: Int, slow: Boolean) {
        require(name.isNotBlank()) { "observer name is required" }
        require(capacity > 0) { "observer capacity must be positive" }
        cancel(name)
        val observer = ControlledObserver(name, slow)
        synchronized(monitor) {
            observers[name] = observer
        }
        runtime.getObject().fanoutRuntime().observe(Contracts.EVENT_CHANNEL, capacity).subscribe(observer)
    }

    fun release(name: String) {
        synchronized(monitor) {
            observers[name]?.release()
        }
    }

    fun cancel(name: String) {
        synchronized(monitor) {
            observers.remove(name)?.cancel()
        }
    }

    fun waitFor(name: String, timeoutMillis: Long) {
        val deadline = System.nanoTime() + timeoutMillis * 1_000_000L
        synchronized(monitor) {
            while (entries.none { it.observer == name } && System.nanoTime() < deadline) {
                monitor.wait(((deadline - System.nanoTime()) / 1_000_000L).coerceAtLeast(1L).coerceAtMost(250L))
            }
            check(entries.any { it.observer == name }) { "observer $name did not receive a status" }
        }
    }

    fun snapshot(): List<FanoutObserverEntry> = synchronized(monitor) { entries.toList() }

    private inner class ControlledObserver(
        private val name: String,
        private val slow: Boolean,
    ) : Flow.Subscriber<ZLinkObservedStatus<ZLinkFanoutStatus>> {
        private var subscription: Flow.Subscription? = null
        private val released = AtomicBoolean(!slow)

        override fun onSubscribe(value: Flow.Subscription) {
            subscription = value
            value.request(1)
        }

        override fun onNext(value: ZLinkObservedStatus<ZLinkFanoutStatus>) {
            val status = value.status()
            synchronized(monitor) {
                entries += FanoutObserverEntry(
                    name,
                    status.sequence(),
                    status.state().name,
                    status.readyPublisherCount(),
                    value.loss().coalescedCount(),
                    value.loss().discardedTerminalCount(),
                )
                monitor.notifyAll()
                while (!released.get()) {
                    try {
                        monitor.wait(250L)
                    } catch (error: InterruptedException) {
                        Thread.currentThread().interrupt()
                        return
                    }
                }
            }
            subscription?.request(1)
        }

        override fun onError(error: Throwable) {
            synchronized(monitor) {
                monitor.notifyAll()
            }
        }

        override fun onComplete() {
            synchronized(monitor) {
                monitor.notifyAll()
            }
        }

        fun release() {
            released.set(true)
            synchronized(monitor) {
                monitor.notifyAll()
            }
        }

        fun cancel() {
            released.set(true)
            subscription?.cancel()
            synchronized(monitor) {
                monitor.notifyAll()
            }
        }
    }
}
